### ISSUE TITLE
CONTRIBUTING.md : Setting-up Your Fork for Contributions don't works

### ISSUE THREAD

**Issue by Naheulf** [fantaisie-software/purebasic#144]
### Description of the Issue
The steps to setup the fork  for contribution don't works with a new fork and git up to date.

### Steps to Reproduce the Issue

1. Fork fantaisie-software/purebasic from the GitHub website
2. open Git Bash
3. Clone the repo with `git clone https://github.com/<github_username>/purebasic.git`
(Replace `<github_username>` with your own github username.)
4. Move to the repo usin `cd purebasic`
5. Follow steps from [CONTRIBUTING.md#setting-up-your-fork-for-contributions](https://github.com/fantaisie-software/purebasic/blob/c3bca303f9e66a4574b62e775eaf1b78ba4468aa/CONTRIBUTING.md#setting-up-your-fork-for-contributions)

### Expected Behavior
Should works.

### Actual Behavior 1
The second command generate the error: `fatal: 'upstream/devel' is not a commit and a branch 'devel' cannot be created from it`

### Fix 1
Run: `git fetch upstream` after the first command.

### Actual Behavior 2
After use the fix 1 the second command generate another error: `fatal: A branch named 'devel' already exists.`

### Fix 2
On the second command replace the `-b` (lower case B) with `-B` (uppercase B)

### More info
I let more experienced users check if theses fix are good or tricky. And I let english native speaker update the CONTRIBUTING.md file.

**Comment by Naheulf** [comment]
### Description of the Issue
The steps in `Working With devel Branch` also generate errors

### Steps to Reproduce the Issue

1. Follow steps to reproduce from my first message and apply the two fixes
2. Execute the `SYNC`steps from CONTRIBUTING.md

### Expected Behavior
Should not prduce any arror with a fresh install

### Actual Behavior
Command `git pull upstream/devel` generate the following errors:
````
fatal: 'upstream/devel' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
````

### Notes
I don't find fixes for this one.

**Comment by tajmone** [comment]
> The second command generate the error: fatal: 'upstream/devel' is not a commit and a branch 'devel' cannot be created from it

Yes, it seems that the instructions are incomplete (and also partially outdated, since now `devel` has become the default branch the steps for tracking the branch are no longer needed).

The problem is that `upstream` is a remote, not a branch. Your fork will have as a default remote `origin`, so you'll need to manually add the `upstream` remote (pointing to the official repo).

These instructions were added separately, since these steps are only required for contributing, and are not necessary for those who just want to use the repository locally:

https://github.com/fantaisie-software/purebasic/blob/c3bca303f9e66a4574b62e775eaf1b78ba4468aa/CONTRIBUTING.md#setting-up-your-fork-for-contributions

Maybe we should edit the document and make this more clear, by adding a note with a link to the extra steps required for contributing.

Also, it might be worth editing the steps for the `devel` branch, since they probably aren't required any longer (but should add similar notes for `master` branch instead.

**Comment by tajmone** [comment]
Actually, the instructions on how to setup the remote should be added there too, for it's needed to fetch updates, so even non-contributors will need to set up the remote.

I'll try and updated the doc when I find some time. I originally wrote that doc all in one go, so there are probably many things left out.

**Comment by tajmone** [comment]
The main issue that needs to be fixed right now pertains the `devel` branch instructions, since they no longer work (and are not needed) because it was changed into the default branch after the doc was written. But the whole document needs to be revised, and made more clear, possibly also adding simple instruction for those who just wish to clone directly the original repository to use it without contributing, etc.

**Comment by tajmone** [comment]
# Updated CONTRIBUTING via PR

@Naheulf, I've updated the CONTRIBUTING.md document and created a pull request (PR #145). You can preview the changes here:

- https://github.com/fantaisie-software/purebasic/blob/fix-contributing-doc/CONTRIBUTING.md

To be honest, I'm not 100% sure that the fixes are correct. The problem is that the default branch was switched from `master` to `devel`, so I'm assuming that after forking the repository `devel` will also be the default branch on the fork — I'm 99% sure, but can't actually test this since I had forked the repository before the default branch change, and am not able to fork it again.

Furthermore, GitHub soon after switched the default branch name from `master` to `main` (see #112), and for a grace-period implemented some algorithms that will auto-redirect unto `main` any dead references to `master`; so I'm not sure how this might be affecting default branch settings on forks.

The point is that although we can be sure that the `devel` branch is the default one according to Git and the repo settings, we can't entirely exclude that GitHub might change some settings (currently, or in the future) during fork operations.

@Naheulf, when you forked the repository was `devel` alredy the default branch? Is so, after cloning you fork did you find yourself in `devel` or `master`?

**Comment by Naheulf** [comment]
I deleted my fork on github and i restarted all the steps to see if it's works or not.

However i still get some errors:
-  Setting-up Your Fork for Contributions
     - At step 3 `CONFIGURE THE UPSTREAM REMOTE` i got:
        ````
        $ git remote add upstream https://github.com/fantaisie-software/purebasic.git
        fatal: not a git repository (or any of the parent directories): .git
        ````
        To fix it I use the step 4 from my fist message (I change the text to be more explicit): 
        > Move to the local clone using `cd purebasic`
-  Working With `devel` Branch
     - At step 1 `SYNCH` i got the same error as my previous message:
         ````
         $ git pull upstream/devel
         fatal: 'upstream/devel' does not appear to be a git repository
         fatal: Could not read from remote repository.
         ````
         However i found a awful fix:
         ````
         $ git pull ./ upstream/devel
         ````
     - At step 3 `REBASE` i got another error:
         ````
         $ git rebase upstream/devel
         error: cannot rebase: You have unstaged changes.
         error: Please commit or stash them.
         ````
         The obvious way to fix it seem to commit changes. However it's the next step in the guide. Did I forgot something ?
      - At step 5 `PULL-REQUEST` I don't get any errors. However my custom branch is not uploaded to the server.
          I found why. Because my fix for step 3 is the step 4 I not did it again. And since the step 3 leave the local repo on the `devel` branch I neither detect it on step 4 nor send create the commit on the `devel` branch at the same step. So the `push` command say me `Everything up-to-date`.

Notes:
I edited few files between steps 2 `BRANCH` and 3 `REBASE`.
I can't test the rebase command while there no accepted PR on the repo and no conflict files.
You can see the full logs until step 3 `REBASE` at https://gist.github.com/Naheulf/ecbfac962108115b075a5b8659eb5405. (I stop it at this step because I no longer follow `CONTRIBUTING.md` / use GUI tools after)
By the way this answer @tajmone to:
> @Naheulf, when you forked the repository was `devel` already the default branch? Is so, after cloning you fork did you find yourself in `devel` or `master`?

And the answer is: YES

**Comment by tajmone** [comment]
Thanks for spotting the mistakes @Naheulf! I forgot to remove the trailing `.git` from those commands (which is not necessary for cloning, and leads to an error when adding a remote).

I didn't add the `cd purabsic` command because it's already mentioned in the previous step (I'm assuming all these steps are carried out at once, in order, and that end users do have some shell knowledge).

> After fork I'm in devel branch.

Good to know, so GitHub doesn't alter the usual Git settings of the repository. I wasn't sure about that.

-------------------------------------------------------------------------------

The command `git pull upstream/devel` is correct, the only reason why it might fail would be if the `upstream` remote is not setup correctly (the suggested command was failing due to the final `.git`, which should be omitted).

When you type the following command:

```bash
$ git remote -v
```
you should get something like:

```bash
origin  https://github.com/your-user-name/purebasic (fetch)
origin  https://github.com/your-user-name/purebasic (push)
upstream        https://github.com/fantaisie-software/purebasic (fetch)
upstream        https://github.com/fantaisie-software/purebasic (push)
```

where the important part is the `upstreamn` remote being present and pointing to the correct URL (notice that these URLs don't include the ending `git`).


-------------------------------------------------------------------------------

> At step 3 REBASE i got another error:
>
> ```bash
> $ git rebase upstream/devel
> error: cannot rebase: You have unstaged changes.
> error: Please commit or stash them.
> ```
>
> The obvious way to fix it seem to commit changes. However it's the next step in the guide. Did I forgot something ?

At the beginning of the document it clearly states that you should never edit files nor commit on neither `master` nor `devel`. These are upstream branches, so you need to keep them clean at all times in order to be able to pull them locally and use them to rebase your own branches, otherwise they get out of synch.

If you have changes on `devel` (or `master`) you shouldn't commit them, instead you should create and checkout a new branch with a custom name (your changes won't get destroyed), you can then commit them on your new local branch, which you can later on rebase on `devel` after having updated it via `git pull`.

I wouldn't mention any of this in the `CONTRIBUTING.md` document though, for it should be kept simple, else it might be more confusing and scarier than useful. There are enough external links in the document to dig deeper into what the basic operations described therein actually do, and why they are needed. The goal is to provide a reference to start with.

**Comment by Naheulf** [comment]
> The command `git pull upstream/devel` is correct, the only reason why it might fail would be if the `upstream` remote is not setup correctly (the suggested command was failing due to the final `.git`, which should be omitted).
> 
> When you type the following command:
> 
> ```shell
> $ git remote -v
> ```
> 
> you should get something like:
> 
> ```shell
> origin  https://github.com/your-user-name/purebasic (fetch)
> origin  https://github.com/your-user-name/purebasic (push)
> upstream        https://github.com/fantaisie-software/purebasic (fetch)
> upstream        https://github.com/fantaisie-software/purebasic (push)
> ```
> 
> where the important part is the `upstreamn` remote being present and pointing to the correct URL (notice that these URLs don't include the ending `git`).

I first got urls ending with `.git` because I first copy them from the green `code` button available on each project main page to executes commands.
However I also restart all my tests without them. After that I got the same output as you. But I still have the same error.
I also asked for someone to follow theses steps on the french PureBasic discord. @venompb did it and got the same error as me.

<hr>

For the `git rebase upstream/devel` I understood why I got this error. After step 2 performed few files changes. However I didn't run any other git command. And since the command of the step `3. REBASE` is `git checkout devel` it have "moved" all theses changes to the `devel` branch.

> ... you should create and checkout a new branch with a custom name (your changes won't get destroyed), ...

I thought git would keep the changes in my custom branch and reset the files to their devel branch states.

> you can then commit them on your new local branch, which you can later on rebase on `devel` after having updated it via `git pull`.

So, If I made changes to my custom branch I have to commit them BEFORE execute the first command (`git checkout devel`) of the step `3. REBASE`. (That is my error).

<hr>
> I wouldn't mention any of this in the `CONTRIBUTING.md` document though, for it should be kept simple, else it might be more confusing and scarier than useful. There are enough external links in the document to dig deeper into what the basic operations described therein actually do, and why they are needed. The goal is to provide a reference to start with.

I thought this guide was a mini-tutorial on "How to contribute to this repository" and the link was for those who wish to know more about how it works.

**Comment by tajmone** [comment]
> However I also restart all my tests without them. After that, I got the same output as you. But I still have the same error.
I also asked for someone to follow theses steps on the french PureBasic discord. @venompb did it and got the same error as me.

Mh. I must be missing out something then. I'll replay all the steps myself, and try to understand if it might have something to do with custom Git settings, and try to improve that part (it's an important step to get started, after all).

> I thought git would keep the changes in my custom branch and reset the files to their devel branch states.

That's a tricky aspect of Git, one which every Git learned has banged his/her head repeatedly, in puzzlement. As a general rule, Git will not let you destroy uncommitted changes without warning you clearly and loudly, with the exception of some commands which are destructive by nature (e.g. resetting a branch to a specific state) or when you specify the `--force` option. But it's important to understand how Git considers the status of modified files! Changes happen in the working space, and until you stage them you are free to switch branches without loosing them (and if checking out a branch would destroy some changes, Git would warn you and prevent you from doing so). Once you've staged your changes you can't switch branches until you commit them, because these changes are now in the staging area. After committing them they are no longer changes, since they are in the index and the working and staging area are "clean" again (unless you're doing partial commits, to submit your changes into multiple commits, e.g. you've edited a file but stage+committ selected hunks per commit).

> I thought this guide was a mini-tutorial on "How to contribute to this repository" 

More of a quick and dirty way of getting started. There are so many different ways to achieve the same goals in Git that a tutorial would be too long — and there are tons of good tutorials out there. One might argue that a good tutorial should push using SSH instead of HTTP, but these are guidelines to get going, nothing more. 

>  and the link was for those who wish to know more about how it works.

Those links are not optional in real work scenarios. Things do go wrong from time to time, and when this happens with Git there are no shortcuts, one has to grasp what commands do, and which options they provide. Instead of writing a longer tutorial I'd rather wish someone to create an Issue asking for help with a real and practical problem, and help him/her out; after all this is an open source project, so it's all about people interacting with each other in supportive ways. Programmers are strange beasts, at nighttime they do for free what they charge for by the hour during the day, except that they're doing it for people they'll probably never meet in real life. It's all about doing what we like, instead of what others like, and doing it relaxed without deadlines and stress

So, yes, I prefer direct interactions to passive reading, because interactions create bonds among collaborators. Imagine a repository where no one ever interacts beside creating PRs and merging them ... that hardly builds a community around a project, nor it's stimulating. At work is all about getting things done, in hobby projects is more about having fun doing it.

**Comment by tajmone** [comment]
## Error Spotted!

@Naheulf, I've spotted the error, you were right:

> &nbsp;3. **REBASE** — As you're working on your new features, chances are that in the meanwhile the upstream repository might get updated with new pull-requests being merged into `devel`. When this happens, you'll need to _rebase_ your local working branch in order to integrate the new commits that were added to `upstream/devel` since you created your branch (or since your last rebase operation):
>
>     ```bash
>     $ git checkout devel
>     $ git rebase upstream/devel

The first command `git checkout devel` shouldn't be there. Rebasing must be done from within the branch you're working on — i.e. you rebase your branch on `devel`.

I've now fixed that, and also made the text more clear:

- https://github.com/fantaisie-software/purebasic/blob/fix-contributing-doc/CONTRIBUTING.md#working-with-devel-branch

(These are the type of errors that happen when you work late at night...)

Thanks for spotting it.

**Comment by Naheulf** [comment]
> The first command `git checkout devel` shouldn't be there. Rebasing must be done from within the branch you're working on — i.e. you rebase your branch on `devel`, not the other way around (and rebasing `devel` on itself, like in the above instructions, makes no sense in this context since it shouldn't have local changes in the first place).

That makes sens. Now I better understand how you make the rebase.
<hr>

I think I also spot why the `git pull upstream/devel` doesn't work on my computer. The linux man page for this command state:
> git-pull(1) - Linux man page
> 
> Synopsis
> `git pull <options> <repository> <refspec>...`

however if I use `gut pull --help` on windows it show me a local HTML file (D:/Program Files/Git/mingw64/share/doc/git-doc/git-pull.html) with this content:
> git-pull(1) Manual Page
> 
> SYNOPSIS
> `git pull [<options>] [<repository> [<refspec>…​]]`

If I understood it: `<repository>` is the repo that should be updated and `<refspec>` the source repo. However in your `git pull upstream/devel` command the `upstream/devel` part is `<refspec>`. And since git on windows required to set `<repository>` to use `<refspec>`...
And if it's the case I can also assume you use Linux (or mac), right ?

**Comment by tajmone** [comment]
> however if I use gut pull --help on windows it show me a local HTML file

Yes, Git for Windows doesn't include man-pages support, so it falls back on opening HTML docs with the default browser.

Git commands don't differ across different platforms, although some commands on Windows might work slightly different due to OS differences (e.g. the file system being case insensitive, [although Win 10 introduced an option to make selected directories case sensitive], the lack of Unix-like file permissions, etc.).

BTW, I work under Windows too.

# Git's Refspec (the interals)

Git's Refspec is the way Git maps references from remotes to a local repo. It's a rather complex topic for it deals with Git's internals, but for the most part you won't have to worry about it since you'll be using references shorthands in your daily commands.

If you use just `git pull`, Git will pull the currently checkout branch from the default remote (by convention, `origin`). Since we're trying to bring into our local repository the updates from the upstream repository (in order to update both our local copy as well as our fork on GitHub) we'll need to pull them from `upstream` (then push them to `origin`).

Basically, since you can't do these operations directly on GitHub's server, you carry them out on you local machine/repo, which acts as an intermediary — remember, Git is a _distributed_ and non-centralized system, although in practice we rely on GitHub as our central server of reference. But you could also host a copy of the repo on GitLab, Bitbucket, etc. (which is rather common), so there's no real central server, it's just that the official repository by Fred is the one on GitHub.

When you use `git pull upstream/devel` you're telling Git that you want to fetch and merge into the current branch the contents from the `devel` branch hosted at GitHub, under Fred's account (the refspec is what allows Git to map `upstream` to `github.com/fantaisie-software/purebasic/`, which is the base URL reference for all operations on the `upstream` remote). Again, another remote could point to GitLab, and another even to [an anonymous darknet like i2p] via encrypted peer-to-peer decentralized protocols, etc.

# Refspec vs References in Daily Use

But in your daily use, you won't have to worry too much about all the nitty gritty of Git's internals (called "plumbing", and usually manually setting a remote is all you'll have to worry about. Since you'll be working mostly inside a single repository, you won't have to specify a full repository, the remote will suffice (technically speaking, the `fantaisie-software/purebasic`,  `Naheulf/purebasic`, and your local clone are three different repositories).

The point is that Git abstracts aways the "plumbing" by exposing the "porcelain" only in the user interface for common operations (there are plumbing and porcelain commands, you'll rarely use the former). At the porcelain level, the slash notation denotes `<remote>/<branch>`.

## Commits, Branches, Tags

The important aspect to grasp here, regarding references, is that most commands don't work only for branches, you can also checkout a specific commit or a tag. A branch is just a pointer to the latest commit (HEAD) of a named path in the repository's graph (DAG), in other words it's a shorthand pointing to a commit, where the shorthand notation remains the same but the commit it points to "moves" as more commits are added (or deleted). Internally, the only unit that matter are commits, and each commit has a unique hash identifier; the rest (branch names, tags, etc.) are just syntactic sugar to simplify managing commits (i.e. not having to type their hash in every command).

A practical example would be this, when tagged releases will be available on this repository, you'll be able to checkout a specific tag (e.g. `5.73`) to access the repository state as it was at the time of a new release (e.g. PB 5.73). Once you know the version scheme adopted in tag releases, you don't need to know which branch or commit the release is at, you can reference its tag (which is a pointer to a specific commit).

Bare in mind that by using a graphic Git frontend you'll be spared many common tasks, e.g. having to set-up tracking branches, since the GUI will do that for you when you checkout a remote branch which is not yet locally available (it will assume you want to track it from where you checked it out, and push changes to your own `origin` if it comes from another remote).

For more info on the Refspec, see:

- [StackOverflow » Git - What is “Refspec”]
- [Git Pro » §10.5 Git Internals — The Refspec]


-------------------------------------------------------------------------------

By the way, your detailed feedback is very useful for it allows me to see the gaps I've left behind in the CONTRIBUTING guide. When creating a quick guide is always difficult to chose what to include out and what to leave out, and it's easy to give for granted some notions that might not be so obvious for beginners, and which could help better contextualize the instructions. I'll try to add a few more inline links in the various steps, so the reader can find explanatory tutorials right at the end of each step introducing new concepts.



[although Win 10 introduced an option to make selected directories case sensitive]: https://www.howtogeek.com/354220/how-to-enable-case-sensitive-folders-on-windows-10/ "How to Enable Case Sensitive File and Folder Names on Windows 10"

[Git Pro » §10.5 Git Internals — The Refspec]: https://git-scm.com/book/en/v2/Git-Internals-The-Refspec

[StackOverflow » Git - What is “Refspec”]: https://stackoverflow.com/questions/44333437/git-what-is-refspec#44333810

[an anonymous darknet like i2p]: https://geti2p.net/en/blog/post/2020/03/06/git-over-i2p

**Comment by tajmone** [comment]
PS: I've added the angle brackets around `your-branch-name`. You reminded me that Git commands use them in their `--help`, so let's stick to the conventional notation.

**Comment by Naheulf** [comment]
OK. This time I think I really spotted why the pull command don't works on my computer.

The https://git-scm.com/docs/git-pull#_description page write:
> \<repository\> should be the name of a remote repository as passed to [git-fetch[1]](https://git-scm.com/docs/git-fetch). \<refspec\> can name an arbitrary remote ref (for example, the name of a tag) or even a collection of refs with corresponding remote-tracking branches (e.g., refs/heads/*:refs/remotes/origin/*), but usually it is the name of a branch in the remote repository.

And the https://git-scm.com/docs/git-fetch#_configured_remote_tracking_branches page write:
> When `git fetch` is run with explicit branches and/or tags to fetch on the command line, e.g. `git fetch origin master`, the \<refspec\>s given on the command line determine what are to be fetched (e.g. `master` in the example, which is a short-hand for `master:`, which in turn means "fetch the *master* branch but I do not explicitly say what remote-tracking branch to update with it from the command line"), and the example command will fetch *only* the *master* branch. The remote.\<repository\>.fetch values determine which remote-tracking branch, if any, is updated.

And in this case there a space between `origin` and `master`. That mean the branch name (`master` in this case) is the `<refspec>` arg on the command line. 

Considerate that, the commands for the SYNCH step should be: (note the space instead of the slash)
```sh
$ git checkout devel      # change active branch to devel branch on the local repository
$ git pull upstream devel # pull the devel branch on the remote repository on the active branch
```
And THIS works. (I also tested with other branch names and the `-v` arg to check what

Notes:
 - Except when I state it I always use the command line for my test to be as close as possible to the given instructions. So the gui tool can't be a source of issues. For theses tests I only use Tortoise git (my gui tools) for commits and see revision graph, logs and other statistics.
 - I use windows 8.1 on my computer (for some part because windows don't allow me to move to windows 10). So all Win10 changes doesn't apply to me.

**Comment by tajmone** [comment]
> Considerate that, the commands for the SYNCH step should be: (note the space instead of the slash)

You're right! **I'VE FIXED IT!**

The confusing aspect is that many other Git commands do use the slash notation (e.g. checkout), so I must have mixed up commands syntax in the document while editing.

---

# Older Windows Editions and Git...

> I use windows 8.1 on my computer (for some part because windows don't allow me to move to windows 10). So all Win10 changes doesn't apply to me.

That's a pity, you'll be missing out some important Git related new features and tools which Microsoft has been investing on (having acquired GitHub, they have now huge interests in promoting Git). Among other things, you'll be missing features like WSL/WSL2, tools like [Windows Terminal], and the plethora of Linux commands and tools which have been added to Win 10 (cURL, Tar, etc.).

My PC originally came with Win 7, which was freely updated to Win 8, and then MS really pushed hard the free Win 10 update, for over a year. From what I remember, there was a period in which any Window version could be freely updated to Win 10 (OEM versions included, of course).

When I wrote the CONTRIBUTING guide I was assuming that most Windows users would have simply migrated to Win 10 during the free-update period (which was further extended by an extra "grace" period). It's possible that Windows Vista, 7 and 8 might not integrate as well as Win 10 when it comes to Git and the various required tools; I'll need to find out more about this, and maybe update the document accordingly — if you do find any Win 8 specific problems, please do open a dedicated Issue, so that we might mention them in the repo's Wiki (I think that's the right place to provide extra info and links for older OSs, rather than the main repo documents).

> **NOTE** — We should add to the Wiki a list of the Git related limitations for Vista, 7 and 8, and first hand experience from people who are using them with Git would be of great help. 

As time passes, more Linux like features are going to be added to Win 10, and more tools and features to better integrate with Git and GitHub. At some point, the gap between Win 10 and older Win editions might grow beyond what can be worked around, for these features won't be provided for older OSs, and if Git for Windows development will rely on these features being present, it could mean that older OSs might be running Git in reduced functionality at some point, or (worse) being unable to use newer Git releases.

Anyhow, if you decide to switch to Win 10 by buying it, go directly for the Pro version, which has lots of extra developers tools (just having access to the virtual machines features, or full-featured Docker support, is worth the price difference). Next time I need to reinstall Windows I'll be switching to Win 10 Pro myself.



[Windows Terminal]: https://docs.microsoft.com/en-us/windows/terminal/
